packages/protocol/contracts/Median.sol

```diff
@@ -0,0 +1,441 @@
+      _swap(arr, 6, 7);   _swap(arr, 8, 9);  _swap(arr, 10, 11);
+    }
+    else if (arr.length == 15) {
+      _swap(arr, 1, 14);  _swap(arr, 2, 13);  _swap(arr, 3, 12);
+      _swap(arr, 4, 11);  _swap(arr, 5, 10);  _swap(arr, 6, 9);
+      _swap(arr, 7, 8);   _swap(arr, 0, 7);   _swap(arr, 1, 6);
+      _swap(arr, 2, 5);   _swap(arr, 3, 4);   _swap(arr, 9, 14);
+      _swap(arr, 10, 13); _swap(arr, 11, 12); _swap(arr, 0, 3);
+      _swap(arr, 4, 7);   _swap(arr, 8, 11);  _swap(arr, 1, 2);
+      _swap(arr, 5, 6);   _swap(arr, 9, 10);  _swap(arr, 13, 14);
+      _swap(arr, 0, 1);   _swap(arr, 2, 8);   _swap(arr, 10, 11);
+      _swap(arr, 3, 9);   _swap(arr, 4, 5);   _swap(arr, 6, 12);
+      _swap(arr, 7, 13);  _swap(arr, 1, 4);   _swap(arr, 5, 10);
+      _swap(arr, 11, 14); _swap(arr, 2, 3);   _swap(arr, 6, 9);
+      _swap(arr, 12, 13); _swap(arr, 7, 8);   _swap(arr, 1, 2);
+      _swap(arr, 3, 4);   _swap(arr, 5, 8);   _swap(arr, 11, 12);
+      _swap(arr, 13, 14); _swap(arr, 6, 10);  _swap(arr, 7, 9);
+      _swap(arr, 2, 3);   _swap(arr, 4, 6);   _swap(arr, 9, 11);
+      _swap(arr, 12, 13); _swap(arr, 5, 7);   _swap(arr, 8, 10);
+      _swap(arr, 4, 5);   _swap(arr, 6, 7);   _swap(arr, 8, 9);
+      _swap(arr, 10, 11); _swap(arr, 3, 4);   _swap(arr, 5, 6);
+      _swap(arr, 7, 8);   _swap(arr, 9, 10);  _swap(arr, 11, 12);
+    }
+    else {
+      _swap(arr, 0, 15);  _swap(arr, 1, 14);  _swap(arr, 2, 13);
+      _swap(arr, 3, 12);  _swap(arr, 4, 11);  _swap(arr, 5, 10);
+      _swap(arr, 6, 9);   _swap(arr, 7, 8);   _swap(arr, 0, 7);
+      _swap(arr, 1, 6);   _swap(arr, 2, 5);   _swap(arr, 3, 4);
+      _swap(arr, 8, 15);  _swap(arr, 9, 14);  _swap(arr, 10, 13);
+      _swap(arr, 11, 12); _swap(arr, 0, 3);   _swap(arr, 4, 7);
+      _swap(arr, 8, 11);  _swap(arr, 12, 15); _swap(arr, 1, 2);
+      _swap(arr, 5, 6);   _swap(arr, 9, 10);  _swap(arr, 13, 14);
+      _swap(arr, 0, 1);   _swap(arr, 2, 8);   _swap(arr, 10, 11);
+      _swap(arr, 14, 15); _swap(arr, 3, 9);   _swap(arr, 4, 5);
+      _swap(arr, 6, 12);  _swap(arr, 7, 13);  _swap(arr, 1, 4);
+      _swap(arr, 5, 10);  _swap(arr, 11, 14); _swap(arr, 2, 3);
+      _swap(arr, 6, 9);   _swap(arr, 12, 13); _swap(arr, 7, 8);
+      _swap(arr, 1, 2);   _swap(arr, 3, 4);   _swap(arr, 5, 8);
+      _swap(arr, 11, 12); _swap(arr, 13, 14); _swap(arr, 6, 10);
+      _swap(arr, 7, 9);   _swap(arr, 2, 3);   _swap(arr, 4, 6);
+      _swap(arr, 9, 11);  _swap(arr, 12, 13); _swap(arr, 5, 7);
+      _swap(arr, 8, 10);  _swap(arr, 4, 5);   _swap(arr, 6, 7);
+      _swap(arr, 8, 9);   _swap(arr, 10, 11); _swap(arr, 3, 4);
+      _swap(arr, 5, 6);   _swap(arr, 7, 8);   _swap(arr, 9, 10);
+      _swap(arr, 11, 12);
+    }
+    return arr[k];
+  }
+
+  /* Swap two elements of an array iff the first element
+   * is greater than the second.
+   * @param arr an array of unsigned integers
+   * @param i the first index
+   * @param j the second index
+   */
+  function _swap
```

Comment:
The name is misleading. Maybe `conditionalSwap`?

Reply:
good call

Reply:
That's gonna make those above functions uglier -- is the trade-off worth it?

Reply:
I think so because I never realized that those were not regular swaps before this review. Maybe `conSwap()` or something like that.